sdks/java/harness/src/main/java/org/apache/beam/fn/harness/debug/DataSampler.java

```diff
@@ -0,0 +1,134 @@
+   * @param sampleEveryN Sets how often to sample.
+   */
+  public DataSampler(int maxSamples, int sampleEveryN) {
+    this.maxSamples = maxSamples;
+    this.sampleEveryN = sampleEveryN;
+  }
+
+  // Maximum number of elements in buffer.
+  private final int maxSamples;
+
+  // Sampling rate.
+  private final int sampleEveryN;
+
+  // The fully-qualified type is: Map[PCollectionId, OutputSampler]. In order to sample
+  // on a PCollection-basis and not per-bundle, this keeps track of shared samples between states.
+  private final Map<String, OutputSampler<?>> outputSamplers = new ConcurrentHashMap<>();
+
+  /**
+   * Creates and returns a class to sample the given PCollection in the given
+   * ProcessBundleDescriptor. Uses the given coder encode samples as bytes when responding to a
+   * SampleDataRequest.
+   *
+   * @param pcollectionId The PCollection to take intermittent samples from.
+   * @param coder The coder associated with the PCollection. Coder may be from a nested context.
+   * @param <T> The type of element contained in the PCollection.
+   * @return the OutputSampler corresponding to the unique PBD and PCollection.
+   */
+  public <T> OutputSampler<T> sampleOutput(String pcollectionId, Coder<T> coder) {
+    outputSamplers.putIfAbsent(
+        pcollectionId, new OutputSampler<>(coder, this.maxSamples, this.sampleEveryN));
+    return (OutputSampler<T>) outputSamplers.get(pcollectionId);
+  }
+
+  /**
+   * Returns all collected samples. Thread-safe.
+   *
+   * @param request The instruction request from the FnApi. Filters based on the given
+   *     SampleDataRequest.
+   * @return Returns all collected samples.
+   */
+  public BeamFnApi.InstructionResponse.Builder handleDataSampleRequest(
+      BeamFnApi.InstructionRequest request) {
+    BeamFnApi.SampleDataRequest sampleDataRequest = request.getSampleData();
+
+    Map<String, List<byte[]>> responseSamples =
+        samplesFor(sampleDataRequest.getPcollectionIdsList());
+
+    BeamFnApi.SampleDataResponse.Builder response = BeamFnApi.SampleDataResponse.newBuilder();
+    for (String pcollectionId : responseSamples.keySet()) {
+      ElementList.Builder elementList = ElementList.newBuilder();
+      for (byte[] sample : responseSamples.get(pcollectionId)) {
+        elementList.addElements(
+            SampledElement.newBuilder().setElement(ByteString.copyFrom(sample)).build());
+      }
+      response.putElementSamples(pcollectionId, elementList.build());
+    }
+
+    return BeamFnApi.InstructionResponse.newBuilder().setSampleData(response);
+  }
+
+  /**
+   * Returns a map from PCollection to its samples. Samples are filtered on
+   * ProcessBundleDescriptorIds and PCollections. Thread-safe.
+   *
+   * @param pcollections Filters all PCollections on this set. If empty, allows all PCollections.
+   * @return a map from PCollection to its samples.
+   */
+  private Map<String, List<byte[]>> samplesFor(List<String> pcollections) {
```

Comment:
You could inline this method above which would make it clear that we only cover this logic when synchronized